### DISCUSSION TITLE
Default directives for multiple `md` files in a directory

### DISCUSSION THREAD

**Discussion by mpf** [marp-team/marp#409] [Q&A] [answered]
I have many `md` slides in a directory, and each has exactly the same yaml frontmatter that I copy into it each time I create a new file:
```
---
marp: true
headingDivider: 2
paginate: true
---
```
Is there a way to have default frontmatter for every `md` file in a directory, say, by setting up a single directives file? I'd prefer this approach, because I can then centrally maintain the options for all of the slides in that directory.

**Comment by yhatt** [comment, marked as answer]
No. Our tools have no support to supply default directive values from the outside of Markdown file, _for keeping consistent reproducibility in a single markdown file._

**Reply by mpf** [reply]
Is there a way to set directives through a custom `css` file?

**Reply by yhatt** [reply]
No.

An only escape hatch is to write [Marpit plugin](https://marpit-api.marp.app/), for setting default directive values after [the process of parsing directives](https://github.com/marp-team/marpit/blob/main/src/markdown/directives/parse.js).

```javascript
// engine.js
const marpitDefaultDirectivesPlugin =
  ({ global: defaultGlobalDirectives, local: defaultLocalDirectives }) =>
  (md) => {
    md.core.ruler.after(
      'marpit_directives_global_parse',
      'marpit_apply_default_global_directives',
      (state) => {
        if (state.inlineMode) return

        md.marpit.lastGlobalDirectives = {
          ...defaultGlobalDirectives,
          ...md.marpit.lastGlobalDirectives,
        }
      }
    )

    md.core.ruler.after(
      'marpit_directives_parse',
      'marpit_apply_default_local_directives',
      (state) => {
        if (state.inlineMode) return

        for (const token of state.tokens) {
          if (token.meta && token.meta.marpitDirectives) {
            token.meta.marpitDirectives = {
              ...defaultLocalDirectives,
              ...defaultGlobalDirectives,
              ...token.meta.marpitDirectives,
            }
          }
        }
      }
    )
  }

module.exports = ({ marp }) =>
  marp.use(
    marpitDefaultDirectivesPlugin({
      global: {
        headingDivider: 2,
      },
      local: {
        paginate: true,
      },
    })
  )
```

This `engine.js` is available in [Marp CLI](https://github.com/marp-team/marp-cli).

```
marp --engine engine.js ./your-markdown-directory
```

**Reply by mpf** [reply]
Wonderful! many thanks, @yhatt.
